AdminWebsite/AdminWebsite/Security/CachedUserClaimBuilder.cs

```diff
@@ -0,0 +1,34 @@
+using AdminWebsite.Helper;
+using AdminWebsite.Services;
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace AdminWebsite.Security
+{
+    public interface ICachedUserClaimBuilder
+    {
+        Task<IEnumerable<Claim>> BuildAsync(string username, string cacheKey);
+    }
+
+    public class CachedUserClaimBuilder : ICachedUserClaimBuilder
+    {
+        private readonly IClaimsCacheProvider _claimsCacheProvider;
+        private readonly IUserAccountService _userAccountService;
+
+        public CachedUserClaimBuilder(IClaimsCacheProvider claimsCacheProvider, IUserAccountService userAccountService)
+        {
+            _claimsCacheProvider = claimsCacheProvider;
+            _userAccountService = userAccountService;
+        }
+
+        public async Task<IEnumerable<Claim>> BuildAsync(string username, string cacheKey)
+        {
+            return await _claimsCacheProvider
+            .GetOrAdd(cacheKey, async key => new AdministratorRoleClaimsHelper
```

Comment:
Feels like this code should be somewhere outside the cache. This is the functionality of how you actually get the user profile, not just how you cache it?

Reply:
Yeh, lets do it.. But where ?